File: Data Structures & Algorithms/queue/submission-1.py
class ListNode:
    def __init__(self, val, next=None, prev=None):
        self.val = val
        self.next = next
        self.prev = prev

class Deque:
    def __init__(self):
        self.head = ListNode(-1)
        self.tail = ListNode(-1, prev=self.head)
        self.head.next = self.tail

    def isEmpty(self) -> bool:
        return self.head.next == self.tail

    def append(self, value: int) -> None:
        node = ListNode(value, next=self.tail)
        tail = self.tail
        tail.prev.next = node
        node.prev = tail.prev
        tail.prev = node

    def appendleft(self, value: int) -> None:
        node = ListNode(value, prev=self.head)
        head = self.head
        head.next.prev = node
        node.next = head.next
        head.next = node

    def pop(self) -> int:
        if self.isEmpty(): return -1
        tail = self.tail
        poped = tail.prev
        poped.prev.next = tail
        tail.prev = poped.prev
        return poped.val

    def popleft(self) -> int:
        if self.isEmpty(): return -1
        head = self.head
        poped = head.next
        head.next = poped.next
        poped.next.prev = head
        
        return poped.val

File: Data Structures & Algorithms/design-linked-list/submission-1.kt
class MyLinkedList() {

    private class ListNode(val value: Int, var next: ListNode? = null, var prev: ListNode? = null)

    private var head: ListNode? = null
    private var tail: ListNode? = null
    private var size: Int = 0

    fun get(index: Int): Int {
        val node = getNode(index)
        return node?.value ?: -1
    }

    fun addAtHead(`val`: Int) {
        val newNode = ListNode(`val`, next = head)
        if (head == null) {
            tail = newNode
        } else {
            head?.prev = newNode
        }
        head = newNode
        size++
    }

    fun addAtTail(`val`: Int) {
        val newNode = ListNode(`val`, prev = tail)
        if (tail == null) {
            head = newNode
        } else {
            tail?.next = newNode
        }
        tail = newNode
        size++
    }

    fun addAtIndex(index: Int, `val`: Int) {
        if (index !in 0..size) return

        if (index == 0) {
            addAtHead(`val`)
            return
        }

        if (index == size) {
            addAtTail(`val`)
            return
        }

        val current = getNode(index) ?: return
        val previous = current.prev
        val newNode = ListNode(`val`, next = current, prev = previous)

        previous?.next = newNode
        current.prev = newNode

        size++
    }

    fun deleteAtIndex(index: Int) {
        val node = getNode(index) ?: return

        val previous = node.prev
        val next = node.next

        if (previous == null) {
            head = next
        } else {
            previous.next = next
        }

        if (next == null) {
            tail = previous
        } else {
            next.prev = previous
        }

        size--
    }

    private fun getNode(index: Int): ListNode? {
        if (index !in 0..<size) return null

        var curIndex = 0
        var curNode = head

        while (curIndex < index) {
            curNode = curNode?.next
            curIndex++
        }
        
        return curNode
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * var obj = MyLinkedList()
 * var param_1 = obj.get(index)
 * obj.addAtHead(`val`)
 * obj.addAtTail(`val`)
 * obj.addAtIndex(index,`val`)
 * obj.deleteAtIndex(index)
 */
